src/storage/src/hummock/recent_filter.rs

```diff
@@ -12,44 +12,100 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::borrow::Borrow;
 use std::collections::{HashSet, VecDeque};
 use std::fmt::Debug;
 use std::hash::Hash;
+use std::sync::LazyLock;
 use std::time::{Duration, Instant};
 
 use parking_lot::RwLock;
+use prometheus::core::{AtomicU64, GenericCounter, GenericCounterVec};
+use prometheus::{
+    register_int_counter_vec_with_registry, register_int_gauge_with_registry, IntGauge, Registry,
+};
+use risingwave_common::monitor::GLOBAL_METRICS_REGISTRY;
 
-pub trait RecentFilterKey = Eq + Hash + Send + Sync + Debug + Clone + 'static;
+pub static GLOBAL_RECENT_FILTER_METRICS: LazyLock<RecentFilterMetrics> =
+    LazyLock::new(|| RecentFilterMetrics::new(&GLOBAL_METRICS_REGISTRY));
+
+pub struct RecentFilterMetrics {
+    pub recent_filter_items: IntGauge,
+    pub recent_filter_ops: GenericCounterVec<AtomicU64>,
+    pub recent_filter_inserts: GenericCounter<AtomicU64>,
+    pub recent_filter_queries: GenericCounter<AtomicU64>,
+}
+
+impl RecentFilterMetrics {
+    fn new(registry: &Registry) -> Self {
+        let recent_filter_items = register_int_gauge_with_registry!(
+            "recent_filter_items",
+            "Item numbers of the recent filter.",
+            registry
+        )
+        .unwrap();
+
+        let recent_filter_ops = register_int_counter_vec_with_registry!(
+            "recent_filter_ops",
+            "Ops of the recent filter.",
+            &["op"],
+            registry
+        )
+        .unwrap();
+
+        let recent_filter_inserts = recent_filter_ops.with_label_values(&["insert"]);
+        let recent_filter_queries = recent_filter_ops.with_label_values(&["query"]);
+
+        Self {
+            recent_filter_items,
+            recent_filter_ops,
+            recent_filter_inserts,
+            recent_filter_queries,
+        }
+    }
+}
 
 #[derive(Debug)]
-pub struct RecentFilter<K>
-where
-    K: RecentFilterKey,
-{
+pub struct RecentFilter<K> {
     refresh_interval: Duration,
-    inner: RwLock<CacheRefillFilterInner<K>>,
+    inner: RwLock<RecentFilterInner<K>>,
 }
 
 #[derive(Debug)]
-struct CacheRefillFilterInner<K>
-where
-    K: RecentFilterKey,
-{
+struct RecentFilterInner<K> {
     last_refresh: Instant,
     layers: VecDeque<RwLock<HashSet<K>>>,
 }
 
-impl<K> RecentFilter<K>
-where
-    K: RecentFilterKey,
-{
+impl<K> RecentFilterInner<K> {
+    fn try_rotate(&mut self, refresh_interval: &Duration) {
+        if &self.last_refresh.elapsed() > refresh_interval {
+            if let Some(removed) = self.layers.pop_front() {
+                GLOBAL_RECENT_FILTER_METRICS
+                    .recent_filter_items
+                    .add(0 - removed.read().len() as i64);
```

Comment:
It looks a little bit weird to add a negative value here.

Reply:
Oh, there is `sub`. I was looking for `dec_by()`